src/Phinx/Db/Adapter/AbstractDirectActionTrait.php

```diff
@@ -0,0 +1,423 @@
+     * @return AlterInstructions
+     */
+    abstract protected function getDropForeignKeyInstructions($tableName, $constraint);
+
+    /**
+     * Returns the instructions to drop the specified foreign key from a database table.
+     *
+     * @param string $tableName The table where the foreign key constraint is
+     * @param array $columns The list of column names
+     * @return AlterInstructions
+     */
+    abstract protected function getDropForeignKeyByColumnsInstructions($tableName, $columns);
+
+    /**
+     * {@inheritdoc}
+     */
+    public function dropTable($tableName)
+    {
+        $instructions = $this->getDropTableInstructions($tableName);
+        $this->executeAlterSteps($tableName, $instructions);
+    }
+
+    /**
+     * Returns the instructions to drop the specified database table.
+     *
+     * @param string $tableName Table Name
+     * @return AlterInstructions
+     */
+    abstract protected function getDropTableInstructions($tableName);
+
+    /**
+     * {@inheritdoc}
+     */
+    public function renameTable($tableName, $newTableName)
+    {
+        $instructions = $this->getRenameTableInstructions($tableName, $newTableName);
+        $this->executeAlterSteps($tableName, $instructions);
+    }
+
+    /**
+     * Returns the instructions to rename the specified database table.
+     *
+     * @param string $tableName Table Name
+     * @param string $newTableName New Name
+     * @return AlterInstructions
+     */
+    abstract protected function getRenameTableInstructions($tableName, $newTableName);
+
+    /**
+     * {@inheritdoc}
+     */
+    public function changePrimaryKey(Table $table, $newColumns)
+    {
+        $instructions = $this->getChangePrimaryKeyInstructions($table, $newColumns);
+        $this->executeAlterSteps($table->getName(), $instructions);
+    }
+
+    /**
+     * Returns the instructions to change the primary key for the specified database table.
+     *
+     * @param Table $table Table
+     * @param string|array|null $newColumns Column name(s) to belong to the primary key, or null to drop the key
+     * @return AlterInstructions
+     */
+    abstract protected function getChangePrimaryKeyInstructions(Table $table, $newColumns);
+
+    /**
+     * {@inheritdoc}
+     */
+    public function changeComment(Table $table, $newComment)
+    {
+        $instructions = $this->getChangeCommentInstructions($table, $newComment);
+        $this->executeAlterSteps($table->getName(), $instructions);
+    }
+
+    /**
+     * Returns the instruction to change the comment for the specified database table.
+     *
+     * @param Table $table Table
+     * @param string|null $newComment New comment string, or null to drop the comment
+     * @return AlterInstructions
+     */
+    abstract protected function getChangeCommentInstructions(Table $table, $newComment);
+
+    /**
+     * {@inheritdoc}
+     */
+    public function executeActions(Table $table, array $actions)
```

Comment:
I would prefer if this function lived in another file. Seems to be very misplaced here

Reply:
Perhaps a `DirectPlanExecutorTrait` file? I wasn't happy with its inclusion, either.

Reply:
What’s the problem with having it in the PDO adapter?

Reply:
The plan executor relies on the underlying implementations of the direct actions throughout. It is, in essence, a plan executor based on a series of direct actions.

Reply:
No, it is calling `executeAlterSteps` which is different from Direct actions in that the alter steps usually comes from a planner, and are not executed by hand. That's why I don't think those correspond to `DirectActionsInterface`.

An alternative is to make another interface having `executeAlterSteps` and friends and making DirectActions extend it.